hmachine/src/main/java/org/hortonmachine/hmachine/modules/statistics/kerneldensity/OmsStructuralSimilarity.java

```diff
@@ -0,0 +1,409 @@
+package org.hortonmachine.hmachine.modules.statistics.kerneldensity;
+
+import oms3.annotations.Execute;
+import org.geotools.coverage.grid.GridCoverage2D;
+import org.hortonmachine.gears.libs.exceptions.ModelsIllegalargumentException;
+import org.hortonmachine.gears.libs.exceptions.ModelsRuntimeException;
+import org.hortonmachine.gears.libs.modules.HMModel;
+import org.hortonmachine.gears.libs.modules.HMRaster;
+import org.jaitools.media.jai.kernel.KernelFactory;
+
+import javax.media.jai.KernelJAI;
+import java.io.IOException;
+import java.util.stream.IntStream;
+
+public class OmsStructuralSimilarity extends HMModel {
+
+
+    public GridCoverage2D inMap1 = null;
+
+    public GridCoverage2D inMap2 = null;
+
+    public double pK1 = 0.01;
+    public double pK2 = 0.03;
+
+    public double pRelevanceMean = 1;
+    public double pRelevanceVariance = 1;
+    public double pRelevancePattern = 1;
+
+    public int pKernel = 3;
+
+    public int pRadius = 10;
+
+    public boolean doConstant = false;
+
+    public GridCoverage2D outStructuralSimilarity = null;
+
+    public GridCoverage2D outMeanSimilarity = null;
+
+    public GridCoverage2D outVarianceSimilarity = null;
+
+    public GridCoverage2D outPatternSimilarity = null;
+
+
+    private volatile boolean errorOccurred = false;
+    private volatile String errorMessage;
+
+
+    @Execute
+    public void process() throws Exception {
+
+        GridCoverage2D meanMap1 = windowMean(inMap1);
```

Comment:
I see you tend to use gridcoverages and convert them to HMRasters once you need them.

It is best that you create an HMRaster at the very begin of the module for all the maps you need and then pass these to the various window* methods.

Then at the end of the modeul you build the output GridCoverages from the HMRaster as you need.

Reply:
Hi @moovida thanks for the comments! I'll implement your suggestions later in the morning

Reply:
I implemented the changes. Now all the window functions take HMRasters as arguments and also return HMRasters instead of a GridCoverage2D. I don't know if you see an issue with that. My thought process is that as soon a GridCoverage that will be certainly used is produced I convert it to HMRaster

Reply:
Yes @diegomvd , apart of the return from the methods, which I commented, it looks fine).
Your thought was ok, but the conversion between from GC to HMRaster has a cost. So it is best to use HMRaster (which has all the methods you need) and then convert ony at the end.